pkg/hive/feature_lifecycle.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package hive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+
+	"github.com/cilium/hive/cell"
+	"golang.org/x/exp/maps"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type FeatureLifecycleInterface interface {
+	Append(Feature, cell.Hook) error
+	Start(Feature, context.Context, *slog.Logger) error
+	Stop(Feature, context.Context, *slog.Logger) error
+
+	IsRunning(Feature) bool
+	List() []Feature
+}
+
+type Feature string
+type FeatureLifecycle struct {
+	mu     lock.Mutex
+	hooks  map[Feature][]cell.Hook
+	status map[Feature]bool
+}
+
+func NewFeatureLifecycle() *FeatureLifecycle {
+	return &FeatureLifecycle{
+		hooks:  make(map[Feature][]cell.Hook),
+		status: make(map[Feature]bool),
+	}
+}
+
+// Append adds a hook to the feature hooks, marking the feature as not running.
+// It returns an error if the feature is already running.
+func (fl *FeatureLifecycle) Append(f Feature, h cell.Hook) error {
+	fl.mu.Lock()
+	defer fl.mu.Unlock()
+
+	if status, ok := fl.status[f]; ok && status {
+		return fmt.Errorf("cannot add hooks to a running feature: %s", f)
+	}
+
+	fl.hooks[f] = append(fl.hooks[f], h)
+	fl.status[f] = false
+
+	return nil
+}
+
+// Start attempts to start a feature by executing its associated hooks.
+// It returns an error if the feature is already running
+// or if any hook fails to start.
+func (fl *FeatureLifecycle) Start(f Feature, c context.Context, l *slog.Logger) error {
+	fl.mu.Lock()
+	defer fl.mu.Unlock()
+
+	if status, ok := fl.status[f]; ok && status {
+		return fmt.Errorf("feature %s is already running", f)
+	}
+
+	l.Debug("Starting hooks for", "feature", f)
+	for _, hook := range fl.hooks[f] {
+		if err := hook.Start(c); err != nil {
+			l.Error("Start hook failed", "error", err)
```

Comment:
When a startup failure occurs, you currently may leave half of your cells in a started state, the normal lifecycle shuts down any already started cells. You might consider doing the same

Reply:
The FeatureLifecycle will be managed by some new logic (a new LifecycleManager) that will ensure the proper start/stop in case of errors, so the new logic should handle the corner cases in which the start / stop fails. As far I've seen the DefaultLifecycle do not stop the cells rather the hive does, so that's why I think the new LifecycleManager should manage this error case.